pkg/controller/elasticsearch/driver/upscale_state.go

```diff
@@ -101,41 +122,102 @@ func (s *upscaleState) recordMasterNodeCreation() {
 	if s.isBootstrapped {
 		s.allowMasterCreation = false
 	}
+	s.recordNodesCreation(1)
 }
 
-func (s *upscaleState) canAddMasterNode() bool {
-	return s.allowMasterCreation
+func (s *upscaleState) canCreateMasterNode() bool {
+	return s.getMaxNodesToCreate(1) == 1 && s.allowMasterCreation
 }
 
-func (s *upscaleState) limitMasterNodesCreation(
-	actualStatefulSets sset.StatefulSetList,
-	ssetToApply appsv1.StatefulSet,
-) appsv1.StatefulSet {
-	if !label.IsMasterNodeSet(ssetToApply) {
-		return ssetToApply
+func (s *upscaleState) recordNodesCreation(count int32) {
+	s.recordedCreates += count
+}
+
+func (s *upscaleState) getMaxNodesToCreate(noMoreThan int32) int32 {
+	if s.createsAllowed == nil {
+		// unbounded, so allow all that was requested
+		return noMoreThan
 	}
 
-	targetReplicas := sset.GetReplicas(ssetToApply)
-	actual, alreadyExists := actualStatefulSets.GetByName(ssetToApply.Name)
-	actualReplicas := int32(0)
-	if alreadyExists {
-		actualReplicas = sset.GetReplicas(actual)
+	left := *s.createsAllowed - s.recordedCreates
+	if left < noMoreThan {
+		return left
+	}
+
+	return noMoreThan
+}
+
+func (s *upscaleState) limitNodesCreation(
+	actual appsv1.StatefulSet,
+	toApply appsv1.StatefulSet,
+) (appsv1.StatefulSet, error) {
+	if label.IsMasterNodeSet(toApply) {
+		return s.limitMasterNodesCreation(actual, toApply)
+	}
+
+	actualReplicas := sset.GetReplicas(actual)
+	targetReplicas := sset.GetReplicas(toApply)
+	if actualReplicas == targetReplicas {
+		return toApply, nil
+	}
+
+	if err := s.buildOnce(); err != nil {
+		return appsv1.StatefulSet{}, err
+	}
+	nodespec.UpdateReplicas(&toApply, common.Int32(actualReplicas))
```

Comment:
Can we move that to the `else` branch? To make the flow easier to understand (either we increase, either we keep the same).

Reply:
You mean the rest of the func after L166? Not sure if this improves readability for me, but not really strong opinion, so it's done :) Let me know if you meant something else.

Reply:
After moving code around this is no longer relevant.